feat(plh_facilitator): shared parent group functionality

jfmcquade opened the pull request:
PR Checklist

- [x] PR title descriptive (can be used in release notes)
- [x] Merge after #2899  (currently targets PR branch for clearer diffs)

## Description

Adds a dedicated PLH Parent Group service to manage shared parent groups functionality. This functionality is specifically required for the `plh_facilitator_*` deployments.

This leverages the shared data feature introduced in #2899. It does not make use of the realtime updates functionality, instead opting to expose actions to manually author triggers for the relevant functionality.

Exposes the following template actions to authoring:

| Action | Params | Description |
|---|---|---|
| `plh_parent_group: push` | `parent_group_id`: local ID of the parent group to share (optional), `parent_groups_data_list`: name of local data list storing parent groups data, `parents_data_list`: name of local data list storing parents data | Pushes updates to shared data. If a `parent_group_id` is provided, this data is updated, otherwise data for all previously shared parent groups is updated |
| `plh_parent_group: add_cofacilitator` | `parent_group_id`: local ID of the parent group to share, `auth_id`: auth ID of the co-facilitator with whom to share the parent group, `parent_groups_data_list`: name of local data list storing parent groups data, `parents_data_list`: name of local data list storing parents data | Pushes/updates the targeted parent group to shared data (or updates if already shared), and adds the specified co-facilitator ID as a member of that shared data entity |
| `plh_parent_group: remove_cofacilitator` | `parent_group_id`: local ID of the parent group to update, `auth_id`: auth ID of the co-facilitator to remove, `parent_groups_data_list`: name of local data list storing parent groups data, `parents_data_list`: name of local data list storing parents data | Updates the targeted parent group in shared data to remove the specified co-facilitator ID as a member of that shared data entity |
| `plh_parent_group: pull` | `parent_group_id`: local ID of the parent group to pull data for (optional), `parent_groups_data_list`: name of local data list storing parent groups data, `parents_data_list`: name of local data list storing parents data, `completion_tracking_data_list`: name of the local data list used for tracking module completion (optional) | Pulls updates from shared data. If a `parent_group_id` is provided, this data is updated, otherwise data for all shared parent groups for which the logged in user has access are updated. If `completion_tracking_data_list` is passed, this data list will be updated with a new field named `completed_${parentGroupId}` |
| `plh_parent_group: create_random` | `parent_group_id`: local ID of the parent group to pull data for (optional), `parent_groups_data_list`: name of local data list storing parent groups data, `parents_data_list`: name of local data list storing parents data | Generates a new local parent group with random details. Intended for use only to demonstrate and debug functionality without implementing a full system for adding parent groups |

[Firebase debug firestore](https://console.firebase.google.com/u/1/project/debug-e8934/firestore/databases/-default-/data/~2Fshared_data~2Fd9acb864-125b-42b1-9a93-712df6807032)

Current parent groups data list: [parent_groups](https://docs.google.com/spreadsheets/d/1px_o58qaL5TIezNKuQf1pnCmboLXymmbDX8xM_EtSv0/edit?gid=2087179638#gid=2087179638)
Current parents data list: parents [parents](https://docs.google.com/spreadsheets/d/1vnZGZLoj7jrkDSR8V8aTCSASH53ucMN6jF1AyN6ksoM/edit?gid=614037394#gid=614037394)

## TODO

Possible follow-ups:
- Support multiple co-facilitators
- Remove parent groups from shared data (e.g. clean up when a co-facilitator is removed)
- Pull action for specific parent group after having access revoked
	- currently throws a firebase permissions error, which could be handled to trigger cleanup of local data

## Dev notes

- Parent groups data list and parents data list could be extracted out to the config level, as opposed to being passed to each action. However, to avoid deployment-specific data in the config I've opted to pass the values through each time to deployment-specific actions

## Git Issues

Relevant: #2858 

## Screenshots/Videos

Demo of functionality using [debug_shared_parent_group](https://docs.google.com/spreadsheets/d/1fSMNZSWxBRoH6WQCrT-xdd5kOJjpM-wW5RQM59uSGgQ/edit?gid=99275306#gid=99275306) template in Chrome windows with two different associated Google accounts. The functionality is a bit obscure, but hopefully should make sense to @ChrisMarsh82 at least:

https://github.com/user-attachments/assets/955875a6-7dd6-4474-83f8-c0df065e57ee

chrismclarke reviewed (changes requested):
Overall I think looking pretty good, I haven't dived too deep into the service code, but I think the fact you are relying mostly on single server queries means you're likely to avoid any of the pitfalls that sometimes comes with managing firestore query billing.

The PR does introduce a few new patterns regarding how we enable deployment-specific features to register things like new action namespaces. I appreciate we don't want to block this PR with trying to come up with a perfect solution, but I have added a number of inline comments which I hope should provide slightly better encapsulation of the deployment-specific knock-ons.

But otherwise I would say once addressed, and if the service code is passing functional tests, all looks like it should be in a good state to merge (after the base PR)

chrismclarke commented:
@jfmcquade 
#2899 merged so I've updated this PR to target `master`

jfmcquade commented on src/app/feature/shared-data/providers/base.ts:
```diff
@@ -7,7 +7,7 @@ export interface SharedDataQueryParams {
    * or entire collection (e.g. "group_1/messages")
    **/
-  id: string;
+  id?: string;
```
Yes, that's the use-case I was trying to implement. See [line 200 here](https://github.com/IDEMSInternational/open-app-builder/pull/2910/files#diff-4bd37b5e366ab5768cb5fe6be9ba0b6722de528c4e76788dec8fa39f7a2eefaeR200): I want a query that returns all documents that the logged-in user has access to. We can't specify the docs because there could be new docs that have been shared with the user that they don't know about yet (i.e. they don't have any local record of). Is there a better way to do this?

chrismclarke replied:
That makes sense. Yeah it's a bit confusing because I'm using the term `id` but when used with firestore it actually refers to the path to any collection or subcollection. For now this isn't strictly required as there is a default `shared_data` collection that is used to store all data, but in the future we might want to support top-level named collections such as `shared_parent_groups`, where we would want the user to pass a `parent_groups` id (with the shared_ inferred).

There is also some added confusion as I've only defined one set of sharedDataQueryParams for both single and multiple queries, when they could be subtly different.

But for now I think the best solution is to keep the id as required (as it is strictly required for the querySingle method), but in your service pass an empty string when querying multiple and wanting the top-level collection. You'll see how the firebase provider code follows a similar pattern itself

```ts
  public queryMultiple$(params: SharedDataQueryParams) {
    const { id = "" } = params;
    const resourcePath = id ? `${COLLECTION}/${id}` : COLLECTION;
    const collectionRef = collection(this.db, resourcePath).withConverter(sharedDataConverter);

    // Retrieve docs where public true or use included
    const docsQuery = this.buildDocumentQuery(collectionRef, params);

    // const privateDocs = query(collectionRef, where());
    return this.queryToObservable<ISharedDataCollection>(docsQuery);
  }
  ```
  
  But yes all of this could do with a little bit of tidying at some point

jfmcquade replied:
Thanks, updated with https://github.com/IDEMSInternational/open-app-builder/pull/2910/commits/5d771bae803d0d5e52439dafaa868861ffd8a030

jfmcquade commented:
Thanks @chrismclarke, I've restructured the plh-specific code as suggested. See also my reply to the unresolved inline comment.

@ChrisMarsh82 has been testing with the deployment and things generally seem to be working as expected

ChrisMarsh82 commented:
@jfmcquade Parents aren't being transferred. It could be that I have blanks in the columns. Not every field in the parents is always populated.

chrismclarke commented:
> Thanks @chrismclarke, I've restructured the plh-specific code as suggested. See also my reply to the unresolved inline comment.

Thanks, looking tidier for sure - I'm not sure if perhaps the changes to actions didn't get pushed yet (just to import from the plh module instead of directly setting `plh_parent_group` in the flowtype), but I think the rest is looking good to me. 

See suggestion to resolve the only other outstanding issue in comment

chrismclarke reviewed (changes requested):
>Thanks @chrismclarke, I've restructured the plh-specific code as suggested. See also my reply to the unresolved inline comment.

Thanks, looking tidier for sure - I'm not sure if perhaps the changes to actions didn't get pushed yet (just to import from the plh module instead of directly setting plh_parent_group in the flowtype), but I think the rest is looking good to me.

See suggestion to resolve the only other outstanding issue in comment

jfmcquade commented:
> I'm not sure if perhaps the changes to actions didn't get pushed yet (just to import from the plh module instead of directly setting plh_parent_group in the flowtype), but I think the rest is looking good to me.

Ah yeh, looks like I overlooked that one. Updated with https://github.com/IDEMSInternational/open-app-builder/pull/2910/commits/133895815d27ee68e8a829a5af84888743bf26c5. It ended up being a bit fiddly to avoid a circular import (between `flowTypes.ts` and and the plh components module (`packages/components/plh/index.ts`), so I added an intermediary file, `packages/components/plh/actions.ts`, to export the action names from.

chrismclarke commented:
> > I'm not sure if perhaps the changes to actions didn't get pushed yet (just to import from the plh module instead of directly setting plh_parent_group in the flowtype), but I think the rest is looking good to me.
> 
> Ah yeh, looks like I overlooked that one. Updated with [1338958](https://github.com/IDEMSInternational/open-app-builder/commit/133895815d27ee68e8a829a5af84888743bf26c5). It ended up being a bit fiddly to avoid a circular import (between `flowTypes.ts` and and the plh components module (`packages/components/plh/index.ts`), so I added an intermediary file, `packages/components/plh/actions.ts`, to export the action names from.
> 
> UPDATE: In fact, this was still causing issues, see [failed test run](https://github.com/IDEMSInternational/open-app-builder/actions/runs/14773619656/job/41477873001?pr=2910). So for now I've just kept the plh actions definition hardcoded in flowTypes, see [66a4d10](https://github.com/IDEMSInternational/open-app-builder/commit/66a4d10a95a83d9f882ffa84e647db8aab56da1e).

Thanks for checking, yeah you're right, the flowtypes is meant to be independent, so shouldn't be importing anything from a deployment repo. A better pattern would likely revolve around simply including the type when calling the action registry, e.g. `templateActionRegistry.register<IPlhAction>` which passes through necessary types for type-safe registry (in this case knowing that the feature should only assign to the `plh_parent_group` namespace). Although then we would also need similar method to concatenate both core and extended actions when calling things like action triggers, or writing spec tests that reference rows. So I think for now what you've done is a reasonable compromise.

chrismclarke reviewed (approved):
Code looks good to me, I'll leave it to @ChrisMarsh82 for functional testing
